### ISSUE TITLE
Layouts Combinations Display Bug

### ISSUE THREAD

**Issue by sifran-github** [freeplane/freeplane#1075]
**Describe the bug**
Some layouts combinations will create an issue with the map display. For example, views with children that have a "side layout".

**Example - Screenshots**
![image](https://user-images.githubusercontent.com/5730355/229485960-a2c8816c-6695-4dfa-8510-bfa02efa00f9.png)

**To Reproduce**
1. Look at the attached map
2. Change only the layout of the topic called "TOPIC. CHANGE MY LAYOUT" to any other layout (see table below).
3. You can see the issues.

Summary of the problematic layouts:
![image](https://user-images.githubusercontent.com/5730355/229485275-0ef05fcf-9a37-4054-a82c-e085ed9de97c.png)

Red = still problematic in pre28
Yellow = was problematic in pre24 but got fixed already in pre28

Note: this is a summary only for this specific node. Other mirrored layouts (right to left, top to bottom) might have the opposite layouts as problematic (or maybe other issues).

**Expected behavior**

- Nodes should not be on top of each other.
- The "promise" that the parent node ("TOPIC. CHANGE MY LAYOUT") is giving to its children should not break.

**Mind map files**
[bug_layouts.zip](https://github.com/freeplane/freeplane/files/11137353/bug_layouts.zip)

**Version**
freeplane-1.11.1-pre28 + freeplane-1.11.1-pre24

**Comment by sifran-github** [comment]
It also occurs when the children ("SUB TOPIC") are in other layouts, not just "side layout".

I guess in any layout that the grandchildren ("aaa") are "pushing" the children ("SUB TOPIC") down, then these children are not positioned correctly with the parent ("TOPIC. CHANGE MY LAYOUT").

And I guess the solution should be generic, because... I can also build a map where the grandchildren of grandchildren of grandchildren will have some layout that will "push" their parents/grandparents and increase the space that is taken by this branch.

**Comment by dpolivaev** [comment]
I do not see any problems with layouts like

<img width="148" alt="ok-layouts" src="https://user-images.githubusercontent.com/1558052/229560637-7595e868-8786-4c92-8b48-f88e0b4b975e.png">

Could you please upload screen shots or attach map where I can see them?

**Comment by sifran-github** [comment]
for example for this layout:
![image](https://user-images.githubusercontent.com/5730355/229581640-2112f3f3-004a-430c-a32a-3002cfd6033f.png)
(edit: updated the layout pic)

this is the result in pre28:
![image](https://user-images.githubusercontent.com/5730355/229580693-e171faba-cf94-41b0-a5c5-af21b0c68b4f.png)

In this layout, the tree under this layout should extend to the top direction only, and the bottom-most child should stay in the same position, next to his parent.

You can see that the position of the most bottom child ("SUB TOPIC") is not at the same level as his parent, like the layout suggests. it is pushed down.

The other layouts have the same issue.

Basically, in any layout, if you focus on [1] the parent and [2] the child that supposed to be close to it, you will see the issues more clearly.

**Comment by dpolivaev** [comment]
Preview 1.11.1_29

**Comment by sifran-github** [comment]
Tested Preview 1.11.1_29.
It looks really good. Most of the issues got fixed, with expected results that matche the selected layout.

There are still unexpected results in the following layouts:
![image](https://user-images.githubusercontent.com/5730355/231294443-d12edc0e-e733-49bf-a474-59d082a4e922.png)

for example, in this layout:
![image](https://user-images.githubusercontent.com/5730355/231294537-256ffc43-734d-4b24-a997-ebe9bb8dbb2f.png)
here is the result. I added orange lines & text to emphasis the expected result:
![image](https://user-images.githubusercontent.com/5730355/231294626-8596d391-e43f-4268-bcf8-e4d002ae0f05.png)
as you can see, the entire "box area" of the children is not centered with the parent.

To get some idea on how it should behave, I took an older Freeplane implementation (1.10.x) and looked at how it behaves on a side orientation:
![image](https://user-images.githubusercontent.com/5730355/231294867-648224fe-58d4-49ec-ab63-c4ebcb4e06ff.png)
even if you minimize the children, it stays centered:
![image](https://user-images.githubusercontent.com/5730355/231295774-0f11985a-06dd-4c56-8b02-ea2b312ddba5.png)

Note: in the beginning I thought that the "box area" should include only the immediate children, and not grandchildren, but after I looked at how old Freeplane behaves, I think its safe to assume the "box area" should include the entire tree under the current node.

**Comment by dpolivaev** [comment]
Both issues are not bugs.

**Comment by dpolivaev** [comment]
Freeplane was never able to arbitrarily mix vertical and horizontal layouts.
The user should use combinations that can be laid out  properly

**Comment by dpolivaev** [comment]
On the layouts like https://github.com/freeplane/freeplane/issues/1075#issuecomment-1504156100 ("centered child nodes") folded / unfolded node position relatively to the parent node is fix. Freeplane 1.10 deviated from this rule, it was a bug and it is fixed now.

Also, the current behavior is correct.

**Comment by dpolivaev** [comment]
Layout indication was not properly updated after jumping in / out. it is fixed now.

**Comment by sifran-github** [comment]
> Both issues are not bugs.

> Freeplane was never able to arbitrarily mix vertical and horizontal layouts.

i really dont understand these responses. So what you are saying is that i should not suggest anything, because this is a new feature?

This is a bug in my opinion because it is working not as the layout indicates.
The same layout structure on a different direction works differently. so by a reasonable logic, this is how it should behave.

**Comment by macmarrum** [comment]
I also find it unexpected that e.g. childNodeAlignment BY_CENTER moves the branch off center (favoring one side) in some combinations of layouts.

I know that it's the result of the attempt to keep a node being folded/unfolded in the same spot on the screen.

If I were to assign some priorities, I'd put the childNodesAlignment over the fixed folded/unfolded position.
In other words, I'd prefer folded/unfolded nodes to change their position but have e.g. BY_CENTER truly centered (the whole branch).
Otherwise the layouts deliver on their promise only for "single-layout descendants". And this is unexpected.

**Comment by dpolivaev** [comment]
I am not sure I can live with it.

**Comment by dpolivaev** [comment]
For me, smoothness is more important than symmetry. 
Because I actually never know which form the map has.
I tried to find a middle way to change the behavior for new layouts but keep it for the old ones.

If you say it does not feel consistent, and I agree, then I have no idea how to solve it.
Therefore I am not sure if Freeplane 1.11 can be released this year.

**Comment by macmarrum** [comment]
In v1.10.x, BY_CENTER always keeps the symmetry, whether children are folded or unfolded.

Since v1.11.1, BY_CENTER keeps the symmetry when children are folded. It also keeps the symmetry for unfolded children, but only if all descendants use BY_CENTER and the same orientation.
The lack of symmetry in other conditions feels inconsistent, because it's different from v1.10.x. If I were to answer whether it's critical to keep consistency with v1.10.x, I'd say no. The new approach is unexpected but not unacceptable.

This is similar to v1.11.1 allowing a combination of childNodesLayouts that lead to node overlapping. When compared to v1.10.x, it's for sure unexpected, because the automatic-layout system of v1.10.x prevents node overlapping. But the flexibility gained with the new childNodesLayouts in v1.11.1 provides far more benefits, therefore the new behavior (overlapping nodes) is acceptable.

I vote for going on with the release and keeping the attempts to figure out how to solve it.

**Comment by dpolivaev** [comment]
> In v1.10.x, BY_CENTER always keeps the symmetry, whether children are folded or unfolded.

Yes, and for me it is a bug that I fixed in 1.11. Freeplane was never acting this way before 1.10.

**Comment by sifran-github** [comment]
> > In v1.10.x, BY_CENTER always keeps the symmetry, whether children are folded or unfolded.
> 
> Yes, and for me it is a bug that I fixed in 1.11. Freeplane was never acting this way before 1.10.

I dont understand what do you mean by that.
Take a look at version 1.8.8 for example. The parent stays centered whether or not the children are folded or not:

![image](https://user-images.githubusercontent.com/5730355/232256073-4f45a868-5a4e-46dc-9f0d-5675e17714dc.png)

The way that i see it, in any version (1.8.x, 1.10.x, 1.11.1 pre29), the only the layout that does not keep symmetry (in a layout that suggest symmetry) is the one i mentioned at https://github.com/freeplane/freeplane/issues/1075#issuecomment-1504156100

**Comment by dpolivaev** [comment]
I am going 

1. to change the center layouts according to your suggestions, 
2. to introduce new layout variants which should behave like current center layouts
3. to use icons like 
<img width="113" alt="smooth" src="https://user-images.githubusercontent.com/1558052/232343790-04fed7f6-3202-4205-b3ff-377fabf1869e.png">
with for the new layouts

4. To make the new layouts the defaults

I still do not have a good name instead of "center" for the new layouts. As a variant, I think of "flow" or "flowing" (toptobottom_bothsides_flow).

**Comment by sifran-github** [comment]
Take a look at https://github.com/freeplane/freeplane/discussions/1101. I really believe that it will solve a lot of issues and allow much more control and configuration.

**Comment by dpolivaev** [comment]
I have changed the "centered" layouts so that they are consistent with the rest and added "natural flow" layouts which are better e.g. if you change the layout by moving nodes extra up and down.

preview 1.11.1_30

I don't want to start another round of ideas. I want to release. So if you don't find show stoppers I would release it soon.

**Comment by macmarrum** [comment]
Nice. I like that there is "natural flow" and "centered".

Here's my feedback as a Feature Request
- https://github.com/freeplane/freeplane/issues/1109

**Comment by sifran-github** [comment]
I checked freeplane-1.11.1-pre31.

I can see that the central positioning that i mentioned here is fixed: https://github.com/freeplane/freeplane/issues/1075#issuecomment-1504156100

I don't understand the exact logic of "natural flow", but its not related to this bug, and having several options is good.
I consider this bug to be closed.

**Comment by amariusz** [comment]
Hi all,

I know this is an old issue, but that's the only place where I could find some explanation behind "natural flow" layout:

> I have changed the "centered" layouts so that they are consistent with the rest and added "natural flow" layouts which are better e.g. if you change the layout by moving nodes extra up and down.

I still don't get it though. Is there a simple case in which it would be preferred over centered variant?

Thanks

**Comment by amariusz** [comment]
..OK, I think I can see the difference now. In `centered` mode, the tree (parent) is centered, but when collapsing/expanding branches, the current node moves. In `flow` mode, the tree is not always centered, but when collapsing/expanding the current node, it always stays in the same position.

So I understand that the 'flow' mode should be used when the jumping during collapsing/expanding becomes problematic.

**Comment by amariusz** [comment]
@dpolivaev thanks for confirming

I think "natural flow" layout could also be named "steady folding" or "fixed folding". But I get the logic behind "natural flow".

Generally speaking, these layouts are a fantastic feature. Thank you all so much.
